### PR TITLE
Error during Caffe model conversion with Python3.8

### PR DESCRIPTION
**Pull request by sadolini**
Root cause analysis: Caffe returns error during model parsing with Python 3.8. Looks like root cause is in protobuf implementation for Python3.8

Solution: Add one more possible reason for Caffe error with recommendation to use Python with cpp implementation of protobuf. Also added similar MxNet error to FAQ.

Ticket: 57299

Code:
* [x]  Comments
* [x]  Code style (PEP8)
* [x]  Transformation generates reshape-able IR **N/A**
* [x]  Transformation preserves original framework node names **N/A**

Validation:
* [x]  Unit tests **N/A**
* [x]  Framework operation tests **N/A**
* [x]  Transformation tests **N/A**
* [x]  Model Optimizer IR Reader check **N/A**

Documentation:
* [x]  Supported frameworks operations list **N/A**
* [x]  Guide on how to convert the **public** model **N/A**
* [x]  User guide update **N/A**

### CODE REVIEW

File: docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md [hunk 1]
```diff
@@ -627,4 +627,15 @@ It means that you trying to convert the topology which contains '_contrib_box_nm
   });
 </script>
 
-\endhtmlonly
\ No newline at end of file
+\endhtmlonly
+
+#### 103. What does the message "ModelOptimizer is not able to parse *.caffemodel" means? <a name="question-103"></a>
+
+If file '*.caffemodel' exist and correct and among possible reasons you can see third point about protobuf implementation then you have used python protobuf implementation. In some cases it gives error during model parsing. Then you should build cpp implementation of protobuf by yourself or use one of wheels from ModelOptimizer package. For complete instructions about building `protobuf` from sources, see the appropriate section in [Converting a Model to Intermediate Representation](Config_Model_Optimizer.md).
```
Comment (popovaan) [inline docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md:634]: If file *.caffemodel exist -> If file *.caffemodel exists

**Review by popovaan — APPROVED** [review]

**Review by generalova-kate — APPROVED** [review]

File: model-optimizer/mo/front/caffe/loader.py [hunk 2]
```diff
@@ -130,10 +130,17 @@ def load_caffe_proto_model(caffe_pb2, proto_path: str, model_path: [str, None] =
...
+                          ' in this configuration. Please, use Python version with existing cpp implementation of ' + \
+                          'protobuf library or build it by yourself\n'
         log.error('Exception message: {}\n\n'.format(e) +
                   '    Possible reasons:\n' +
                   '      1. {} does not exist\n'.format(model_path) +
-                  '      2. {} does not have a valid structure\n'.format(model_path), extra={'framework_error': True})
+                  '      2. {} does not have a valid structure\n'.format(model_path) + third_point +
+                  refer_to_faq_msg(103),
+                  extra={'framework_error': True})
         raise FrameworkError('Model Optimizer is not able to parse {}'.format(model_path)) from e
```
Comment (rkazants) [inline model-optimizer/mo/front/caffe/loader.py:145]: why don't you refer error #104?

File: model-optimizer/mo/front/caffe/loader.py [hunk 3]
```diff
@@ -130,10 +130,17 @@ def load_caffe_proto_model(caffe_pb2, proto_path: str, model_path: [str, None] =
                 map = mmap.mmap(infile.fileno(), 0, access=mmap.ACCESS_READ)
                 model.MergeFromString(map)
     except Exception as e:
+        third_point = ''
+        if api_implementation._implementation_type == 'python':
+            third_point = '      3. Python protobuf implementation was used. Some models can\'t be converted ' + \
+                          ' in this configuration. Please, use Python version with existing cpp implementation of ' + \
+                          'protobuf library or build it by yourself\n'
```
Comment (rkazants) [inline model-optimizer/mo/front/caffe/loader.py:137]: I think it is better to move concatentation with `refer_to_faq_msg(103)` here.

File: docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md [hunk 4]
```diff
@@ -627,4 +627,15 @@ It means that you trying to convert the topology which contains '_contrib_box_nm
   });
 </script>
 
-\endhtmlonly
\ No newline at end of file
+\endhtmlonly
+
+#### 103. What does the message "ModelOptimizer is not able to parse *.caffemodel" means? <a name="question-103"></a>
+
```
Comment (rkazants) [inline docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md:633]: also, please double check grammar correctness of sentences.

File: model-optimizer/mo/front/caffe/loader.py (hunk 3 above)
Comment (sadolini) [reply model-optimizer/mo/front/caffe/loader.py:137]: Agree

File: model-optimizer/mo/front/caffe/loader.py (hunk 2 above)
Comment (sadolini) [reply model-optimizer/mo/front/caffe/loader.py:145]: It is MxNet error, from framework itself. We are not catching it in MO.

File: docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md [hunk 5]
```diff
@@ -627,4 +627,15 @@ It means that you trying to convert the topology which contains '_contrib_box_nm
   });
 </script>
 
-\endhtmlonly
\ No newline at end of file
+\endhtmlonly
+
+#### 103. What does the message "ModelOptimizer is not able to parse *.caffemodel" means? <a name="question-103"></a>
```
Comment (rkazants) [inline docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md:632]: means -> mean

File: docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md [hunk 6]
```diff
@@ -627,4 +627,15 @@ It means that you trying to convert the topology which contains '_contrib_box_nm
...
+
+If file '*.caffemodel' exists and is correct and among possible reasons you can see the third point about protobuf implementation then you are using python protobuf implementation. In some cases it gives error during model parsing: for example "'utf-8' codec can't decode byte 0xe0 in position 4: invalid continuation byte in field: mo_caffe.SpatialTransformerParameter.transform_type". You can use Python3.6/3.7 or you can build 'cpp' implementation of protobuf by yourself for your version of Python. For complete instructions about building `protobuf` from sources, see the appropriate section in [Converting a Model to Intermediate Representation](Config_Model_Optimizer.md).
+
+#### 104. What does the message "SyntaxError: 'yield' inside list comprehension" during MxNet\* model conversion means? <a name="question-104"></a>
```
Comment (rkazants) [inline docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md:636]: means -> mean

File: docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md [hunk 7]
```diff
@@ -627,4 +627,15 @@ It means that you trying to convert the topology which contains '_contrib_box_nm
...
+
+You can face with "SyntaxError: 'yield' inside list comprehension" issue during converting MxNet\* models (mobilefacedet-v1-mxnet, brain-tumor-segmentation-0001) on Windows* platform with Python* 3.8 environment. This issue existed due to API changes for `yeld expression` in Python 3.8
```
Comment (rkazants) [inline docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md:638]: yeld -> yield

Comment (generalova-kate) [reply docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md:638]:
```suggestion
You can face with "SyntaxError: 'yield' inside list comprehension" issue during converting MxNet\* models (mobilefacedet-v1-mxnet, brain-tumor-segmentation-0001) on Windows* platform with Python* 3.8 environment. This issue existed due to API changes for `yield expression` in Python 3.8 
```

**Review by rkazants — APPROVED** [review]

File: docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md [hunk 8]
```diff
@@ -627,4 +627,15 @@ It means that you trying to convert the topology which contains '_contrib_box_nm
   });
 </script>
 
-\endhtmlonly
\ No newline at end of file
+\endhtmlonly
+
+#### 103. What does the message "ModelOptimizer is not able to parse *.caffemodel" means? <a name="question-103"></a>
+
+If file '*.caffemodel' exists and is correct and among possible reasons you can see the third point about protobuf implementation then you are using python protobuf implementation. In some cases it gives error during model parsing: for example "'utf-8' codec can't decode byte 0xe0 in position 4: invalid continuation byte in field: mo_caffe.SpatialTransformerParameter.transform_type". You can use Python3.6/3.7 or you can build 'cpp' implementation of protobuf by yourself for your version of Python. For complete instructions about building `protobuf` from sources, see the appropriate section in [Converting a Model to Intermediate Representation](Config_Model_Optimizer.md).
```
[the remark this answers is not in the document]
Comment (generalova-kate) [reply docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md]:
```suggestion
You can face Caffe model conversion error (for example densenet-201 Caffe model on Windows* platform with Python* 3.8 environment). It gives an error during model parsing: "'utf-8' codec can't decode byte 0xe0 in position 4: invalid continuation byte in field: mo_caffe.SpatialTransformerParameter.transform_type". In addition, the error message can complain about protobuf implementation when you are using python protobuf implementation. We suggest some workarounds:
1. You can use Python3.6/3.7 to convert a Caffe model on Windows
2. You can build 'cpp' implementation of protobuf by yourself for your version of Python. For complete instructions about building `protobuf` from sources, see the appropriate section in [Converting a Model to Intermediate Representation](Config_Model_Optimizer.md).
```

File: docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md [hunk 9]
```diff
@@ -627,4 +627,15 @@ It means that you trying to convert the topology which contains '_contrib_box_nm
   });
 </script>
 
-\endhtmlonly
\ No newline at end of file
+\endhtmlonly
+
+#### 103. What does the message "ModelOptimizer is not able to parse *.caffemodel" mean? <a name="question-103"></a>
+
+If a file '*.caffemodel' exists and it is correct, you can observe error due to use of python protobuf implementation. In some cases it gives error during model parsing: for example "'utf-8' codec can't decode byte 0xe0 in position 4: invalid continuation byte in field: mo_caffe.SpatialTransformerParameter.transform_type". You can use Python3.6/3.7 or you can build 'cpp' implementation of protobuf by yourself for your version of Python. For complete instructions about building `protobuf` from sources, see the appropriate section in [Converting a Model to Intermediate Representation](Config_Model_Optimizer.md).
```
Comment (tsavina) [inline docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md:634]:
```suggestion
If a '*.caffemodel' file exists and it is correct, the error possibly occured due to the use of Python protobuf implementation. In some cases, it shows error message during model parsing, for example: "'utf-8' codec can't decode byte 0xe0 in position 4: invalid continuation byte in field: mo_caffe.SpatialTransformerParameter.transform_type". You can either use Python 3.6/3.7 or build 'cpp' implementation of protobuf yourself for your version of Python. For the complete instructions about building `protobuf` from sources, see the appropriate section in [Converting a Model to Intermediate Representation](Config_Model_Optimizer.md).
```

File: docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md [hunk 10]
```diff
@@ -627,4 +627,15 @@ It means that you trying to convert the topology which contains '_contrib_box_nm
...
+
+#### 104. What does the message "SyntaxError: 'yield' inside list comprehension" during MxNet\* model conversion mean? <a name="question-104"></a>
+
+You can face with "SyntaxError: 'yield' inside list comprehension" issue during converting MxNet\* models (mobilefacedet-v1-mxnet, brain-tumor-segmentation-0001) on Windows* platform with Python* 3.8 environment. This issue existed due to API changes for `yield expression` in Python 3.8
```
Comment (tsavina) [inline docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md:638]:
```suggestion
The issue "SyntaxError: 'yield' inside list comprehension" might occur during converting MXNet\* models (mobilefacedet-v1-mxnet, brain-tumor-segmentation-0001) on Windows* platform with Python* 3.8 environment. This issue is caused by API changes for `yield expression` in Python 3.8.
```

File: docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md [hunk 11]
```diff
@@ -627,4 +627,15 @@ It means that you trying to convert the topology which contains '_contrib_box_nm
...
+
+#### 104. What does the message "SyntaxError: 'yield' inside list comprehension" during MxNet\* model conversion mean? <a name="question-104"></a>
+
+You can face with "SyntaxError: 'yield' inside list comprehension" issue during converting MxNet\* models (mobilefacedet-v1-mxnet, brain-tumor-segmentation-0001) on Windows* platform with Python* 3.8 environment. This issue existed due to API changes for `yield expression` in Python 3.8 
+We suggest some workarounds:
+1. Use Python 3.6/3.7 to convert MxNet\* models on Windows
+2. Update mxnet but in that case, you can have conflicts between previously installed PyPI dependencies: pip install mxnet=1.7.0.post2
```
Comment (tsavina) [inline docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md:641]:
```suggestion
2. Update MXNet: pip install mxnet=1.7.0.post2
Note that you might have conflicts between previously installed PyPI dependencies.
```

File: docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md [hunk 12]
```diff
@@ -627,4 +627,15 @@ It means that you trying to convert the topology which contains '_contrib_box_nm
...
+
+If a file '*.caffemodel' exists and it is correct, you can observe error due to use of python protobuf implementation. In some cases it gives error during model parsing: for example "'utf-8' codec can't decode byte 0xe0 in position 4: invalid continuation byte in field: mo_caffe.SpatialTransformerParameter.transform_type". You can use Python3.6/3.7 or you can build 'cpp' implementation of protobuf by yourself for your version of Python. For complete instructions about building `protobuf` from sources, see the appropriate section in [Converting a Model to Intermediate Representation](Config_Model_Optimizer.md).
+
+#### 104. What does the message "SyntaxError: 'yield' inside list comprehension" during MxNet\* model conversion mean? <a name="question-104"></a>
+
+You can face with "SyntaxError: 'yield' inside list comprehension" issue during converting MxNet\* models (mobilefacedet-v1-mxnet, brain-tumor-segmentation-0001) on Windows* platform with Python* 3.8 environment. This issue existed due to API changes for `yield expression` in Python 3.8 
+We suggest some workarounds:
+1. Use Python 3.6/3.7 to convert MxNet\* models on Windows
```
Comment (tsavina) [inline docs/MO_DG/prepare_model/Model_Optimizer_FAQ.md:640]:
```suggestion
1. Use Python 3.6/3.7 to convert MXNet\* models on Windows
```

**Review by tsavina — APPROVED** [review]

**Pull request merged**